sdk/core/azure-core/inc/azure/core/etag.hpp

```diff
@@ -0,0 +1,194 @@
+
+  /**
+   * @brief Represents an HTTP ETag.
+   */
+  class ETag {
+  private:
+    Nullable<std::string> m_value;
+
+  private:
+    /*
+    2.3.2.  Comparison
+
+      There are two entity-tag comparison functions, depending on whether
+      or not the comparison context allows the use of weak validators:
+
+      o  Strong comparison: two entity-tags are equivalent if both are not
+          weak and their opaque-tags match character-by-character.
+
+      o  Weak comparison: two entity-tags are equivalent if their
+          opaque-tags match character-by-character, regardless of either or
+          both being tagged as "weak".
+
+       +--------+--------+-------------------+-----------------+
+       | ETag 1 | ETag 2 | Strong Comparison | Weak Comparison |
+       +--------+--------+-------------------+-----------------+
+       | W/"1"  | W/"1"  | no match          | match           |
+       | W/"1"  | W/"2"  | no match          | no match        |
+       | W/"1"  | "1"    | no match          | match           |
+       | "1"    | "1"    | match             | match           |
+       +--------+--------+-------------------+-----------------+
+
+    // etag:                            //This is possible and means no etag is present
+    // etag:""
+    // etag:"*"                         //This means the etag is value '*'
+    // etag:"some value"                //This means the etag is value 'some value'
+    // etag:/W""                        //Weak eTag
+    // etag:*                           //This is special, means any etag
+    // If-Match header can do this
+    // If-Match:"value1","value2","value3"  // Do this if any of these match
+    //
+
+    */
+    bool ETagComparison(const ETag& left, const ETag& right) const
+    {
+
+      // If both values are empty then we consider the ETag equal
+      if (!left.m_value && !right.m_value)
+        return true;
+
+      if (!left.m_value || !right.m_value)
+        return false;
+
+      // Previous comparison checked that both are Weak or both are strong
+      //  Thuse we can decide which comparison to do based on one of them
+      if (!left.IsWeak())
+      {
+        // Strong ETag so comparison is that the values match character for character
+        return left.m_value.GetValue().compare(right.m_value.GetValue()) == 0;
+      }
+      else
+      {
+        // Left side is weak ETag
+        if (right.IsWeak())
+        {
+          return left.m_value.GetValue().compare(right.m_value.GetValue()) == 0;
+        }
+        // Trim the weak tag
+        auto leftLength = left.m_value.GetValue().length();
+
+        if (leftLength - 2 != right.m_value.GetValue().length())
+        {
+          return false;
+        }
+        return left.m_value.GetValue().compare(2, leftLength - 2, right.m_value.GetValue()) == 0;
+      }
+      return false;
+    }
+
+  protected:
+    /**
+     * @brief Construct an empty (null) #ETag.
+     */
+    ETag() {}
```

Comment:
Why is this constructor protected, i.e., accessible only to derived classes? Also, this should be explicitly defaulted as `ETag() = default;` so folks aren't left wondering why we've gone out of our way to make the default constructor non-trivial.